File: MoviesMVVMAPPWithPagingJava/app/src/main/java/com/masterclass/moviesmvvmappwithpagingjava/adapters/MoviesAdapter.java
package com.masterclass.moviesmvvmappwithpagingjava.adapters;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.RequestManager;
import com.masterclass.moviesmvvmappwithpagingjava.databinding.SingleMovieLayoutBinding;
import com.masterclass.moviesmvvmappwithpagingjava.models.Movie;


public class MoviesAdapter extends PagingDataAdapter<Movie, MoviesAdapter.MovieViewHolder> {

    public static final int LOADING_ITEM = 0;
    public static final int MOVIE_ITEM = 1;
    RequestManager glide;

    public MoviesAdapter(@NonNull DiffUtil.ItemCallback<Movie> diffCallback, RequestManager glide) {
        super(diffCallback);
        this.glide = glide;
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new MovieViewHolder(SingleMovieLayoutBinding.inflate(LayoutInflater.from(parent.getContext()),
                parent,false));
    }

    @SuppressLint("CheckResult")
    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        Movie currentMovie = getItem(position);

        if (currentMovie != null){
            glide.load("https://image.tmdb.org/t/p/w500" + currentMovie.getPosterPath())
                    .into(holder.movieItemBinding.imageViewMovie);
            holder.movieItemBinding.textViewRating.setText(String.valueOf(currentMovie.getVoteAverage()));
        }
    }

    @Override
    public int getItemViewType(int position) {
        return position == getItemCount() ? MOVIE_ITEM : LOADING_ITEM;
    }

    public static class MovieViewHolder extends RecyclerView.ViewHolder{
        SingleMovieLayoutBinding movieItemBinding;
        public MovieViewHolder(@NonNull SingleMovieLayoutBinding movieItemBinding){
            super(movieItemBinding.getRoot());
            this.movieItemBinding = movieItemBinding;
        }
    }
}

File: MoviesMVVMAPPWithPagingKotlin/app/src/main/java/com/masterclass/moviesmvvmappwithpagingkotlin/adapters/MoviesAdapter.kt
package com.masterclass.moviesmvvmappwithpagingkotlin.adapters

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import com.masterclass.moviesmvvmappwithpagingkotlin.databinding.SingleMovieLayoutBinding
import com.masterclass.moviesmvvmappwithpagingkotlin.models.Movie
import com.masterclass.moviesmvvmappwithpagingkotlin.utils.MovieComparator

class MoviesAdapter(private val glide: RequestManager, movieComparator: MovieComparator) :
    PagingDataAdapter<Movie, MoviesAdapter.MovieViewHolder>(MovieDiffCallback()) {

    companion object {
        const val LOADING_ITEM = 0
        const val MOVIE_ITEM = 1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val binding = SingleMovieLayoutBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return MovieViewHolder(binding)
    }

    @SuppressLint("CheckResult")
    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        val currentMovie = getItem(position)
        currentMovie?.let {
            glide.load("https://image.tmdb.org/t/p/w500${it.posterPath}")
                .into(holder.binding.imageViewMovie)
            holder.binding.textViewRating.text = it.voteAverage.toString()
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == itemCount - 1) MOVIE_ITEM else LOADING_ITEM
    }

    class MovieViewHolder(val binding: SingleMovieLayoutBinding) :
        RecyclerView.ViewHolder(binding.root)
}

class MovieDiffCallback : DiffUtil.ItemCallback<Movie>() {
    override fun areItemsTheSame(oldItem: Movie, newItem: Movie): Boolean {
        return oldItem.id == newItem.id
    }

    override fun areContentsTheSame(oldItem: Movie, newItem: Movie): Boolean {
        return oldItem == newItem
    }
}